doc/likwid-sysfeatures.1
.TH LIKWID-SYSFEATURES 1 <DATE> likwid\-VERSION
.SH NAME
likwid-sysfeatures \- retrieve and manipulate hardware and operating system features
.SH SYNOPSIS
.B likwid-sysfeatures
.RB [\-vh]
.RB [ \-V
.IR <verbosity> ]
.RB [ \-a, \-\-all ]
.RB [ \-l, \-\-list ]
.RB [ \-p, \-\-print ]
.RB [ \-O ]
.RB [ \-d, \-\-devices
.IR <device_list> ]
.RB [ \-g, \-\-get
.IR <feature_list> ]
.RB [ \-s, \-\-set
.IR <feature_list> ]


.SH DESCRIPTION
.B likwid-sysfeatures
is a command line application to retrieve and manipulate the state of hardware
and operating system features. It is a common interface for various features
like CPU frequencies, CPU prefetchers and OS-based NUMA balancing. These features
commonly have a topological scope which means that they cover a specific topological
entity. Some exist only once per node. In order to retrieve or manipulate
feature values, all devices have to be specified in the device list. When LIKWID
was built with Nvidia GPU support, likwid-sysfeatures provides access to Nvidia
NVML features.

.SH OPTIONS
.TP
.B \-\^h,\-\-\^help
prints a help message to standard output, then exits.
.TP
.B \-\^v,\-\-\^version
prints version information to standard output, then exits.
.TP
.B \-\^V, \-\-\^verbose <level>
verbose output during execution for debugging. 0 for only errors, 1 for informational output, 2 for detailed output and 3 for developer output
.TP
.B \-\^a,\-\-\^all
list all provided features for the current system and OS
.TP
.B \-\^l,\-\-\^list
list all features with their current value for given devices
.TP
.B \-\^p,\-\-\^print
list available devices for all available device types
.TP
.B \-\^d, \-\-\^devices <device_list>
use devices for listing, getting and setting features
.TP
.B \-\^g, \-\-\^get <feature_list>
Get the values of the specified features of the given devices (
.B <category>.<name>
or just
.B <name>
if unique in a comma-separated list ).
.TP
.B \-\^s, \-\-\^set <feature_list>
Set the values of the specified features on the given devices (
.B <category>.<name>=<value>
or just
.B <name>=<value>
if unique in a comma-separated list ).
.TP
.B \-\^O
Output in RFC-conform CSV

.SH DEVICES
With the
.B \-\^d, \-\-\^devices <device_list>
option, the devices for listing, getting and setting features can be specified.
The syntax is closely related to the syntax used for specifying HW thread lists
for
.B likwid-pin(1)
or
.B likwid-perfctr(1)
but extends it to allow more explicit specification of the topological entities.
All device specifiers with their description:
.RS
.nf
- T : hardware thread
- C : CPU core
- M : NUMA domain
- D : CPU die
- S : CPU socket
- N : Whole node
- GN : Nvidia GPU (if compiled with Nvidia CUDA support)
- GA : AMD GPU (if compiled with AMD ROCm support)
.RE
The general syntax is <specifier>:<comma-separated list or ranges> like M:0,1,4-5.
Multiple device specifiers can be combined with '@' like M:0,1@S:0.

.IP 1. 5
.TP
Specification of hardware threads

.B likwid-sysfeatures --devices T:0,1,4-5 --list

or

.B likwid-sysfeatures --devices 0,1,4-5 --list
.PP
List the current value of all features for the hardware threads with IDs 0, 1, 4
and 5. 'T' is the specifier for hardware threads but can be omitted for convenience. If a
feature is not in hardware threads scope, it will be printed with '-' as value.
.IP 2. 5
.TP
Specification of CPU socket(s)

.B likwid-sysfeatures --devices S:0-1 --list
.PP
List the current value of all features for the topological entity 'CPU socket 0' and 'CPU socket 1'.
.IP 3. 5
.TP
Specification of NUMA domains

.B likwid-sysfeatures --devices M:3,4 --list
.PP
List the current value of all features for the topological entity 'NUMA domain 3' and 'NUMA domain 4'
.IP 4. 5
.TP
Specification of different devices

.B likwid-sysfeatures --devices T:0,1@S:0 --list
.PP
List the current value of all features for hardware threads with ID 0 and 1 as well as 'CPU socket 0'.

.SH EXAMPLES
Some examples for listing the value of features for various devices can be found in the
.B DEVICES
section.
.IP 1. 5
.TP
List all available hardware system and operating system features with information about category, name,
scope, their access mode and a description.

.B likwid-sysfeatures --all
.IP 2. 5
.TP
Get the value of the feature 'prefetch.l2_hwpf' for the hardware thread with ID 0

.B likwid-sysfeatures -d T:0 -g prefetch.l2_hwpf
.IP 3. 5
.TP
Set the value of the feature 'prefetch.l2_hwpf' and 'prefetch.l2_adj_pf' for
the hardware thread with ID 4 to 7 to '0'.

.B likwid-sysfeatures -d T:4-7 -s prefetch.l2_hwpf=0,prefetch.l2_adj_pf=0

.PP
List the current value of all features for the hardware threads with IDs 0, 1, 4
and 5. 'T' is the specifier for hardware threads but can be omitted for convenience. If a
feature is not in hardware threads scope, it will be printed with '-' as value.

.SH AUTHOR
Written by Thomas Gruber <thomas.roehl@googlemail.com> and Michael Panzlaff <michael.panzlaff@fau.de>.
.SH BUGS
Report Bugs on <https://github.com/RRZE-HPC/likwid/issues>.
.SH "SEE ALSO"
likwid-features(1), likwid-powermeter(1), likwid-setFrequencies(1)


CHANGELOG
# Changelog 5.5.0
- Add support for AMD Zen5
- Change hashtable to cwisstable
- New internals for error handling and memory management
- Topdown metrics for Intel SPR, ENR and GNR
- Update for Nvidia GPU support
- Less startup overhead by using optimized Uncore discovery for Intel
- A lot of warnings got fixed
- Wider data types for power readings
- Measure energy consumption for Nvidia GPUs
- Better messages when connection to access daemon fails
- Fix for SPR UPI and M3UPI units
- More peakflops benchmarks in likwid-bench
- Fixes for sysfeatures
- Update to buildsystem to allow parallel builds
- Wrapper mode for GPU monitoring
- Memory controllers (UMC) for AMD Zen4
- Minor fixes for Intel SierraForrest and Coffeelake

# Changelog 5.4.2
- Added support for Intel Emerald Rapids


# Changelog 5.4.1
- Fixes linking errors due to missing bstrlib.h
- Fix for likwid-bench kernel stream_mem
- Fix builds with CUDA for versions 11.2 to 12.6
- Fix sysfeatures with ACCESSMODE=perf_event
- Add AMD Zen1 to Zen3 to sysfeatures
- Add support for apple-cpufreq and cppc driver to sysfeatures
- Fix sysfeatures on ARM architectures

# Changelog 5.4.0
- Support for Intel Granite Rapids (core, energy, uncore)
- Support for Intel Sierra Forrest (core, energy, uncore)
- Support for AMD Bergamo (core, energy, uncore)
- Support for Nvidia Grace (core, uncore)
- Fix: AMD Zen4 DataFabric units
- Fix: Multi-socket RAPL measurements on Sapphire Rapids
- Fix: Energy unit for RAPL DRAM domain for SPR, GNR and SRF
- Fix: Discovery mechanism workaround for UPI and M3UPI units of SPR
- Fix: Intel Westmere Uncore with perf_event backend
- Fix: Fujitsu A64FX (more counters, fixed topology, ...)
- Container bridge to use LIKWID inside container
- Sysfeatures interface reworked with support for various architectures and libraries
- Fix build for NVMON for CUDA 12.6+
- likwid-mpirun: SLURM pinning with cpu_mask feature
- Update of internal hwloc (2.11.2) and Lua (5.4.7) version


# Changelog 5.3.0
- Support for Intel SapphireRapids (Core, Uncore, RAPL)
- Support for AMD Zen4 (Core, Uncore, RAPL)
- Support for Apple M1
- Support for AMD GPUs (MarkerAPI, F90 interface)
- Support for AWS Graviton3 (ARM Neoverse V1)
- Support for HiSilicon TSV110
- Fix of F90 interface installation
- Support for extended umasks in ICX and SPR
- Units for metrics in performance groups
- Library calls to get meta information (version, supported features, etc.)
- Some fixes for direct access mode
- Some fixes for X86 RDPMC detection
- Update of internal hwloc (2.9.3) and Lua (5.4.6) version
- New experimental sysfeatures module

# Changelog 5.2.2
- Add mutex to pinning library
- Fix pin string parsing in pinning library
- Make SBIN path configurable in build system
- Add PKGBUILD for ArchLinux package builds
- Remove accessDaemon double-fork in systemd environements
- Group updates for L2/L3 (mainly AMD Zen)
- Fix multi-initialization in MarkerAPI
- Add energy event scaling for Fujitsu A64FX
- Nvmon: Use Cupti error string to get better warning/error messages
- Nvmon: Store events internally to re-use event strings in stopCounters
- AccessLayer: Catch SIGCHLD to stop sending requests to accessDaemon if it was killed
- likwid-genTopoCfg: Update writing and reading of topology file
- Add INST_RETIRED_NOP event for Intel Icelake (desktop & server)
- Removed some memory leaks
- Improved checks for RDPMC availability
- Add TOPDOWN_SLOTS for perf_event
- Fix for systems with CPU sockets without hwthreads (A64FX FX1000)
- Fix if HOME environment variable is not set (systemd)
- Reader function for perf_event_paranoid in Lua to get state early
- likwid-mpirun: Sanitize np and ppn values to avoid crashes


# Changelog 5.2.1
- Add support for Intel Rocketlake and AMD Zen3 variant (Family 19, Model 0x50)
- Fix for perf_event multiplexing (important!)
- Fix for potential deadlock in MarkerAPI (thx @jenny-cheung)
- Build and runtime fixes for Nvidia GPU backend, updates for CUDA test codes
- peakflops kernel for ARMv8
- Updates for AMD Zen1/2/3 event lists and groups
- Support spaces in MarkerAPI region tags (thx @jrmadsen)
- Use 'online' cpulist instead of 'present'
- Switch CI from Travis-CI to NHR@FAU Cx services
- Document -reset and -ureset for likwid-setFrequencies
- Reset cpuset in unpinned runs
- Remove destructor in frequency module
- Check PID if given through --perfpid
- Intel Icelake: OFFCORE_RESPONSE events
- AccessDaemon: Check PCI init state before using it
- likwid-mpirun: Set mpi type for SLURM automatically
- likwid-mpirun: Fix for skip mask for OpenMPI


# Changelog 5.2.0
- Support for AMD Zen3 (Core + Uncore)
- Support for Intel IcelakeSP (Core + Uncore)
- New affinity code
- Fix for Ivybridge uncore code
- Bypass accessdaemon by using rdpmc instruction on x86_64
- Introduce notion of CPU die in topology module
- Use CPU dies for socket-lock for Intel CascadelakeAP
- Add environment variable LIKWID_IGNORE_CPUSET to break out of current CPUset
- Fixes for affinity module CPUlist sorting
- Build against system-installed hwloc
- Update for Intel SkylakeX/CascadelakeX L3 group
- Rename DataFabric events for all generations of AMD Zen
- Add static cache configuration for Fujitsu A64FX
- Add multiplexing checks for perf_event backend
- Fix for table width of likwid-topology afteradding CPU die column
- Adding RasPi 4 with 32 bit OS as ARMv7
- Add default groups for Intel Icelake desktop
- Fix for likwid-setFrequencies to not apply minFreq when setting governor
- likwid-powermeter: Fix hwthread selection when run with -p
- likwid-setFrequencies: Get measured base frequency if register is not readable
- CLOCK group for all AMD Zen
- Fixes in Nvidia GPU support in NvMarkerAPI and topology module

# Changelog 5.1.1
- Support for Intel Cometlake desktop (Core + Uncore)
- Fix for topology module of Fujitsu A64FX
- Fix for Intel Skylake SP in SNC mode
- Fix for likwid-perfscope
- Fix for CLI argument parsing
- Updated group and data file checkers

src/applications/likwid-sysfeatures.lua
#!<INSTALLED_BINPREFIX>/likwid-lua
--[[
 * =======================================================================================
 *
 *      Filename:  likwid-sysfeatures.lua
 *
 *      Description:  A application to retrieve and manipulate CPU features.
 *
 *      Version:   <VERSION>
 *      Released:  <DATE>
 *
 *      Author:   Thomas Gruber (tr), thomas.roehl@gmail.com
 *      Project:  likwid
 *
 *      Copyright (C) 2016 RRZE, University Erlangen-Nuremberg
 *
 *      This program is free software: you can redistribute it and/or modify it under
 *      the terms of the GNU General Public License as published by the Free Software
 *      Foundation, either version 3 of the License, or (at your option) any later
 *      version.
 *
 *      This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *      WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *      PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 *      You should have received a copy of the GNU General Public License along with
 *      this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 * =======================================================================================
]]
package.path = '<INSTALLED_PREFIX>/share/lua/?.lua;' .. package.path

local likwid = require("likwid")


print_stdout = print
print_stderr = function(...) for k,v in pairs({...}) do io.stderr:write(v .. "\n") end end


local function version()
    print_stdout(string.format("likwid-features -- Version %d.%d.%d (commit: %s)",likwid.version,likwid.release,likwid.minor,likwid.commit))
end

local function example()
    print_stdout("Options:")
end

local function usage()
    version()
    print_stdout("A tool list and modify the states of system/OS features.")
    print_stdout("")
    print_stdout("Options:")
    print_stdout("  -h, --help             Help message")
    print_stdout("  -v, --version          Version information")
    print_stdout("  -a, --all              List all available features")
    print_stdout("  -l, --list             List features and state for given devices")
    print_stdout("  -p, --print            List available devices")
    print_stdout("  -d, --devices <list>   Perform operations on given devices")
    print_stdout("  -g, --get <feature(s)> Get the value of the given feature(s)")
    print_stdout("                         feature format: <category>.<name> or just <name> if unique")
    print_stdout("                         can be a comma-separated list of features")
    print_stdout("  -s, --set <feature(s)> Set feature(s) to the given value")
    print_stdout("                         format: <category>.<name>=<value> or just <name>=<value> if unique")
    print_stdout("                         can be a comma-separated list of features")
    print_stdout("      --saveall <file>   Save all available rw-features to file")
    print_stdout("      --loadall <file>   Load all features from file")
    print_stdout("  -O                     Output results in CSV")
    print_stdout("  -V, --verbose <level>  Set verbosity")
    print_stdout("")
    print_stdout("Examples:")
    print_stdout("  Not all examples shown here may work on your hardware.")
    print_stdout("  List all sysfeature values for Socket 0 and 1.")
    print_stdout("  Please keep in mind that this only prints values applicable to Socket Domain:")
    print_stdout("    $ likwid-sysfeatures --list --devices S:0,1")
    print_stdout("")
    print_stdout("  List available sysfeatures for Cores 0, 1, 4, 5, 6, 7")
    print_stdout("    $ likwid-sysfeatures --list --devices C:0,1,4-7")
    print_stdout("")
    print_stdout("  Get RAPL energy reading for Socket 0:")
    print_stdout("    $ likwid-sysfeatures --get rapl.pkg_energy --devices S:0")
    print_stdout("")
    print_stdout("  Set power limit 1 on Socket 0 to 150W via AMD HSMP:")
    print_stdout("    $ likwid-sysfeatures --set hsmp.pkg_power_limit_cur=150 --devices S:0")
end

local function toGetTable(devList, ft_list)
    -- prepare output table
    local all = {}
    -- first column contains the feature category and name
    local first = {}
    table.insert(first, "Feature/Dev")
    for _,f in pairs(ft_list) do
        table.insert(first, string.format("%s.%s", f.Category, f.Name))
    end
    setmetatable(first, {align = "left"})
    table.insert(all, first)
    -- create one column per given device with the current value of the feature
    for i, dev in pairs(devList) do
        local tab = {}
        table.insert(tab, string.format("%s %s", dev:typeName(), dev:id()))
        for _,f in pairs(ft_list) do
            if dev:typeId() == f.TypeID then
                if f.WriteOnly then
                    table.insert(tab, "(wronly)")
                else
                    local v, err = likwid.sysFeatures_get(string.format("%s.%s", f.Category, f.Name), dev)
                    if v == nil then
                        table.insert(tab, "fail")
                    else
                        table.insert(tab, v)
                    end
                end
            else
                table.insert(tab, "-")
            end
        end
        -- add the hw thread column to the table
        table.insert(all, tab)
    end
    return all
end

if #arg == 0 then
    usage()
    os.exit(0)
end

libinfo = likwid.getLibInfo()
if not libinfo["sysfeaturesSupport"] then
    print_stderr("LIKWID built without sysFeatures support.")
    os.exit(0)
end

-- main variables with defaults
local listFeatures = false
local allFeatures = false
local printDevices = false
local devList = {}
local getList = {}
local setList = {}
local cpuinfo = likwid.getCpuInfo()
local cputopo = likwid.getCpuTopology()
local affinity = likwid.getAffinityInfo()
local verbose = 0
local output_csv = false

-- parse the command line
for opt,arg in likwid.getopt(arg, {"h","v","l","p","d:","g:","s:","a", "O","help","version","list", "print", "set:", "get:","all", "cpus:", "V:", "verbose:", "saveall:", "loadall:"}) do
    if (type(arg) == "string") then
        local s,e = arg:find("-");
        if s == 1 then
            print_stderr(string.format("Argument %s to option -%s starts with invalid character -.", arg, opt))
            print_stderr("Did you forget an argument to an option?")
            os.exit(1)
        end
    end
    if opt == "h" or opt == "help" then
        usage()
        os.exit(0)
    elseif opt == "v" or opt == "version" then
        version()
        os.exit(0)
    elseif opt == "d" or opt == "devices"then
        devList = likwid.createDevicesFromString(arg)
    elseif opt == "l" or opt == "list" then
        listFeatures = true
    elseif opt == "p" or opt == "print" then
        printDevices = true
    elseif opt == "O" then
        output_csv = true
    elseif opt == "V" or opt == "verbose" then
        verbose = tonumber(arg)
    elseif opt == "a" or opt == "all" then
        allFeatures = true
    elseif opt == "g" or opt == "get" then
        getList = likwid.stringsplit(arg, ",")
    elseif opt == "s" or opt == "set" then
        setList = likwid.stringsplit(arg, ",")
    elseif opt == "saveall" then
        saveFeatures = arg
    elseif opt == "loadall" then
        loadFeatures = arg
    elseif opt == "?" then
        print_stderr("Invalid commandline option -"..arg)
        os.exit(1)
    elseif opt == "!" then
        print_stderr("Option requires an argument")
        os.exit(1)
    end
end


-- validate command line input
if (not printDevices) and (not listFeatures) and (not allFeatures) and (not saveFeatures) and (not loadFeatures) and (#getList == 0) and (#setList == 0) then
    print_stderr("No operations specified, exiting...")
    os.exit(1)
end
if (printDevices or listFeatures or allFeatures or saveFeatures or loadFeatures) and (#getList > 0 or #setList > 0) then
    print_stderr("Cannot list features and get/set/load/save at the same time")
    os.exit(1)
end
if #devList == 0 then
    if listFeatures then
        print_stderr("Device selection (-d) required for listing the state of all features")
        os.exit(1)
    elseif #getList > 0 then
        print_stderr("Device selection (-d) required for getting the state of given features")
        os.exit(1)
    elseif #setList > 0 then
        print_stderr("Device selection (-d) required for setting the state of given features")
        os.exit(1)
    end
end

-- set verbosity
if verbose > 0 and verbose <= 3 then
    likwid.setVerbosity(verbose)
end

-- initialize the sysfeatures module
local err = likwid.initSysFeatures()
if err < 0 then
    print_stderr("Cannot initialize HW features module")
    os.exit(1)
end

-- get a list of all features for the system
local ft_list = likwid.sysFeatures_list()

-- print available devices
device_types = {}
device_types[likwid.hwthread] = "HWThread (T)"
device_types[likwid.core] = "Core (C)"
device_types[likwid.numa] = "NUMA (M)"
device_types[likwid.die] = "Die (D)"
device_types[likwid.socket] = "Socket (S)"
device_types[likwid.node] = "Node (N)"
if libinfo["nvidiaSupport"] then
    device_types[likwid.nvidia_gpu] = "Nvidia GPU (GN)"
end
if libinfo["rocmSupport"] then
    device_types[likwid.amd_gpu] = "AMD GPU (GA)"
end

if printDevices then
    for devtype, name in pairs(device_types) do
        print(string.format("%s:", name))
        devices = likwid.getAvailableDevices(devtype)
        if #devices == 0 then
            print("\t<none>")
        else
            print("\t" .. table.concat(devices, ","))
        end
    end
end

-- print the list
if allFeatures then
    local all = {}
    local names = {}
    local types = {}
    local access = {}
    local descs = {}
    local cats = {}
    -- create a table of categories for sorting
    for _,f in pairs(ft_list) do
        found = false
        for _, c in pairs(cats) do
            if c == f.Category then
                found = true
            end
        end
        if not found then
            table.insert(cats, f.Category)
        end
    end
    table.sort(cats)

    -- prepare table for output
    -- first we create tables for the four columns
    table.insert(names, "Feature")
    table.insert(types, "Scope")
    table.insert(access, "Access")
    table.insert(descs, "Description")
    for _,c in pairs(cats) do
        for _,f in pairs(ft_list) do
            if f.Category == c then
                table.insert(names, string.format("%s.%s", f.Category, f.Name))
                table.insert(types, f.Type)
                table.insert(descs, f.Description)
                if f.ReadOnly then
                    table.insert(access, "rdonly")
                elseif f.WriteOnly then
                    table.insert(access, "wronly")
                else
                    table.insert(access, "rw")
                end
            end
        end
    end
    -- add all columns to the table
    setmetatable(names, {align = "left"})
    table.insert(all, names)
    table.insert(all, types)
    table.insert(all, access)
    setmetatable(descs, {align = "left"})
    table.insert(all, descs)

    -- print the table in selected format
    if output_csv then
        likwid.printcsv(all, #all)
    else
        print_stdout("Available features:")
        likwid.printtable(all)
    end

    -- finalize sysfeatures module and exit
    likwid.finalizeSysFeatures()
    os.exit(0)
end

if listFeatures and #devList > 0 then
    -- prepare output table and get all features
    local all = toGetTable(devList, ft_list)

    -- print the table
    if output_csv then
        likwid.printcsv(all, #devList + 1)
    else
        likwid.printtable(all)
    end

    -- finalize sysfeatures module and exit
    likwid.finalizeSysFeatures()
    os.exit(0)
end

if #getList > 0 and #devList > 0 then
    -- filter the full feature list to have only the selected ones
    -- the new list entry contains all entries of the original feature entry plus
    -- the user-given value
    local featList = {}
    for _, f in pairs(getList) do
        for _, l in pairs(ft_list) do
            if f == l.Name or f == string.format("%s.%s", l.Category, l.Name) then
                table.insert(featList, {
                    TypeID = l.TypeID,
                    Type = l.Type,
                    Name = l.Name,
                    Category = l.Category,
                    ReadOnly = l.ReadOnly,
                    WriteOnly = l.WriteOnly,
                })
            end
        end
    end

    -- get all features in the new list
    local tab = toGetTable(devList, featList)
    -- print the table
    if output_csv then
        likwid.printcsv(tab, #devList + 1)
    else
        likwid.printtable(tab)
    end

    -- finalize sysfeatures module and exit
    likwid.finalizeSysFeatures()
    os.exit(0)
end

if #setList > 0 and #devList > 0 then
    -- filter the full feature list to have only the selected ones
    -- the new list entry contains all entries of the original feature entry plus
    -- the user-given value
    local featList = {}
    for _, f in pairs(setList) do
        local t = likwid.stringsplit(f, "=")
        if #t == 2 then
            for _, l in pairs(ft_list) do
                if t[1] == l.Name or t[1] == string.format("%s.%s", l.Category, l.Name) then
                    table.insert(featList, {
                        TypeID = l.TypeID,
                        Type = l.Type,
                        Name = l.Name,
                        Category = l.Category,
                        ReadOnly = l.ReadOnly,
                        WriteOnly = l.WriteOnly,
                        Value = t[2]
                    })
                end
            end
        else
            print_stderr(string.format("Invalid format of '%s' in set list", f))
        end
    end

    -- set all features in the new list
    for i, dev in pairs(devList) do
        local tab = {}
        for _,f in pairs(featList) do
            if verbose > 0 then
                print_stdout(string.format("Setting '%s.%s' to '%s' (Type %s, Resp %d)", f.Category, f.Name, f.Value, f.Type, c))
            end
            local v, err = likwid.sysFeatures_set(f.Name, dev, f.Value)
            if not v then
                print_stderr(string.format("Failed to set feature '%s.%s' to '%s' (Type %s, Resp %s)", f.Category, f.Name, f.Value, f.Type, err))
            end
        end
    end
    -- finalize sysfeatures module and exit
    likwid.finalizeSysFeatures()
    os.exit(0)
end

-- save all read/write features to file
if saveFeatures then
    local file = io.open(saveFeatures, "w")
    -- iterate over all device types
    for devtype, _ in pairs(device_types) do
        -- iterate over all features
        for _,f in pairs(ft_list) do
            local full_name = string.format("%s.%s", f.Category, f.Name)
            -- only allow matching device types and if feature is readable and writable
            if f.TypeID ~= devtype or f.ReadOnly or f.WriteOnly then
                goto next_feat
            end

            -- actually read the features
            for _, dev in pairs(likwid.getAllDevices(devtype)) do
                local lw_dev = likwid.createDevice(devtype, dev)
                local v, err = likwid.sysFeatures_get(full_name, lw_dev)
                if err then
                    print_stderr(string.format("Failed to get feature '%s' on device %s:%d: %s", full_name, lw_dev:typeName(), lw_dev:id(), err))
                    goto next_feat
                end
                file:write(string.format("%s.%s@%s=%s\n", f.Category, f.Name, lw_dev:id(), v))
            end

            ::next_feat::
        end
    end

    file:close()

    likwid.finalizeSysFeatures()
    os.exit(0)
end

-- load all features from file
if loadFeatures then
    for line in io.lines(loadFeatures) do
        -- split string like the following: cpu_freq.governor@5=schedutil
        local part1 = likwid.stringsplit(line, "=")
        if #part1 ~= 2 then
            print_stderr("Invalid line: " .. line)
            os.exit(1)
        end
        local part2 = likwid.stringsplit(part1[1], "@")
        if #part2 ~= 2 then
            print_stderr("Invalid line: " .. line)
            os.exit(1)
        end
        part3 = likwid.stringsplit(part2[1], ".")
        if #part3 ~= 2 then
            print_stderr("Invalid line: " .. line)
            os.exit(1)
        end
        local feat_cat = part3[1]
        local feat_name = part3[2]
        local dev_id = part2[2]
        local value = part1[2]

        local full_name = string.format("%s.%s", feat_cat, feat_name)

        -- get device type of this particular feature
        local devtype = nil
        for _, f in pairs(ft_list) do
            if f.Name == feat_name and f.Category == feat_cat then
                devtype = f.TypeID
                break
            end
        end
        if not devtype then
            print_stderr(string.format("Unknown feature: '%s'", full_name))
            os.exit(1)
        end
        local lw_dev = likwid.createDevice(devtype, dev_id)
        local success, err = likwid.sysFeatures_set(full_name, lw_dev, value)
        if not success then
            print_stderr(string.format("Failed to set feature '%s' on device %s:%d to %s: %s", full_name, lw_dev:typeName(), lw_dev:id(), value, err))
        end
    end

    likwid.finalizeSysFeatures()
    os.exit(0)
end

likwid.finalizeSysFeatures()
os.exit(0)
